File: projects/dejavu.cs/Dejavu.Engine/IrRenderer.cs
using System.Text;
using System.Text.Json.Nodes;
using Dejavu.Language;
using Dejavu.Types;

namespace Dejavu.Engine;

public static class IrRenderer
{
    public static string Render(JsonObject doc, JsonObject ctx)
    {
        var scope = ctx.ToDictionary(p => p.Key, p => p.Value?.DeepClone());
        return RenderNode(doc["body"]!, scope);
    }

    public static string RenderSource(string source, JsonObject ctx)
        => Render(T1Parser.ParseToIr(source), ctx);

    static string RenderNode(JsonNode node, Dictionary<string, JsonNode?> scope)
    {
        var type = node["type"]!.GetValue<string>();
        return type switch
        {
            "Template" => JoinChildren(node["children"]!.AsArray(), scope),
            "Text" => node["value"]!.GetValue<string>(),
            "Comment" => "",
            "Interpolation" => Format(Eval(node["expression"]!, scope), node["raw"]?.GetValue<bool>() == true),
            "Stmt.If" => RenderIf(node, scope),
            "Stmt.For" => RenderFor(node, scope),
            "Stmt.Raw" => node["value"]!.GetValue<string>(),
            "Stmt.Block" => JoinChildren(node["body"]!.AsArray(), scope),
            "Stmt.Extends" or "Stmt.Include" or "Stmt.Super"
                => throw new InvalidOperationException("extends/include/super need a loader"),
            _ => throw new InvalidOperationException($"node not renderable: {type}"),
        };
    }

    static string RenderIf(JsonNode node, Dictionary<string, JsonNode?> scope)
    {
        if (Truthy(Eval(node["test"]!, scope)))
            return JoinChildren(node["consequent"]!.AsArray(), scope);
        foreach (var ei in node["elseIfs"]!.AsArray())
        {
            if (Truthy(Eval(ei!["test"]!, scope)))
                return JoinChildren(ei["consequent"]!.AsArray(), scope);
        }
        if (node["alternate"] is JsonArray alt)
            return JoinChildren(alt, scope);
        return "";
    }

    static string RenderFor(JsonNode node, Dictionary<string, JsonNode?> scope)
    {
        var item = node["item"]!.GetValue<string>();
        var iterable = Eval(node["iterable"]!, scope) as JsonArray
            ?? throw new InvalidOperationException("for iterable must be array");
        var sb = new StringBuilder();
        var indexName = node["index"]?.GetValue<string>();
        for (var i = 0; i < iterable.Count; i++)
        {
            scope.TryGetValue(item, out var prev);
            scope[item] = iterable[i]?.DeepClone();
            JsonNode? prevIdx = null;
            if (indexName is not null)
            {
                scope.TryGetValue(indexName, out prevIdx);
                scope[indexName] = JsonValue.Create(i);
            }
            sb.Append(JoinChildren(node["body"]!.AsArray(), scope));
            if (prev is null) scope.Remove(item); else scope[item] = prev;
            if (indexName is not null)
            {
                if (prevIdx is null) scope.Remove(indexName); else scope[indexName] = prevIdx;
            }
        }
        return sb.ToString();
    }

    static string JoinChildren(JsonArray children, Dictionary<string, JsonNode?> scope)
    {
        var sb = new StringBuilder();
        foreach (var c in children) sb.Append(RenderNode(c!, scope));
        return sb.ToString();
    }

    static string Format(JsonNode? v, bool raw)
    {
        var s = IrFilters.ToString(v);
        return raw ? s : HtmlEscape(s);
    }

    static string HtmlEscape(string s) =>
        s.Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;")
            .Replace("\"", "&quot;").Replace("'", "&#39;");

    static JsonNode? Eval(JsonNode expr, Dictionary<string, JsonNode?> scope)
    {
        var type = expr["type"]!.GetValue<string>();
        return type switch
        {
            "Expr.Literal" => expr["value"]?.DeepClone(),
            "Expr.Identifier" => scope.TryGetValue(expr["name"]!.GetValue<string>(), out var v) ? v?.DeepClone() : null,
            "Expr.Member" => EvalMember(expr, scope),
            "Expr.Index" => EvalIndex(expr, scope),
            "Expr.Binary" => EvalBinary(expr["operator"]!.GetValue<string>(), Eval(expr["left"]!, scope), Eval(expr["right"]!, scope)),
            "Expr.Unary" => EvalUnary(expr["operator"]!.GetValue<string>(), Eval(expr["argument"]!, scope)),
            "Expr.Pipe" => EvalPipe(expr, scope),
            "Expr.Call" => throw new InvalidOperationException("calls not supported in T1 eval"),
            _ => throw new InvalidOperationException($"invalid expression: {type}"),
        };
    }

    static JsonNode? EvalMember(JsonNode expr, Dictionary<string, JsonNode?> scope)
    {
        var obj = Eval(expr["object"]!, scope) as JsonObject;
        var prop = expr["property"]!.GetValue<string>();
        return obj is not null && obj.TryGetPropertyValue(prop, out var v) ? v?.DeepClone() : null;
    }

    static JsonNode? EvalIndex(JsonNode expr, Dictionary<string, JsonNode?> scope)
    {
        var obj = Eval(expr["object"]!, scope);
        var idx = Eval(expr["index"]!, scope);
        if (obj is JsonArray a && idx is JsonValue jv && jv.TryGetValue<int>(out var i))
            return i >= 0 && i < a.Count ? a[i]?.DeepClone() : null;
        if (obj is JsonObject o && idx is JsonValue ks && ks.TryGetValue<string>(out var key))
            return o.TryGetPropertyValue(key, out var v) ? v?.DeepClone() : null;
        return null;
    }

    static JsonNode? EvalPipe(JsonNode expr, Dictionary<string, JsonNode?> scope)
    {
        var val = Eval(expr["expression"]!, scope);
        var args = expr["arguments"]!.AsArray().Select(a => Eval(a!, scope)).ToList();
        return IrFilters.Apply(expr["filter"]!.GetValue<string>(), val, args);
    }

    static JsonNode? EvalUnary(string op, JsonNode? v) => op switch
    {
        "!" => JsonValue.Create(!Truthy(v)),
        "-" when v is JsonValue n && n.TryGetValue<double>(out var d) => JsonValue.Create(-d),
        "+" => v,
        _ => null,
    };

    static JsonNode? EvalBinary(string op, JsonNode? l, JsonNode? r) => op switch
    {
        "+" when AsDouble(l, out var a) && AsDouble(r, out var b) => JsonValue.Create(a + b),
        "+" => JsonValue.Create(IrFilters.ToString(l) + IrFilters.ToString(r)),
        "-" when AsDouble(l, out var a) && AsDouble(r, out var b) => JsonValue.Create(a - b),
        "*" when AsDouble(l, out var a) && AsDouble(r, out var b) => JsonValue.Create(a * b),
        "/" when AsDouble(l, out var a) && AsDouble(r, out var b) => JsonValue.Create(a / b),
        "%" when AsDouble(l, out var a) && AsDouble(r, out var b) => JsonValue.Create(a % b),
        "==" => JsonValue.Create(JsonNode.DeepEquals(l, r)),
        "!=" => JsonValue.Create(!JsonNode.DeepEquals(l, r)),
        "<" when AsDouble(l, out var a) && AsDouble(r, out var b) => JsonValue.Create(a < b),
        "<=" when AsDouble(l, out var a) && AsDouble(r, out var b) => JsonValue.Create(a <= b),
        ">" when AsDouble(l, out var a) && AsDouble(r, out var b) => JsonValue.Create(a > b),
        ">=" when AsDouble(l, out var a) && AsDouble(r, out var b) => JsonValue.Create(a >= b),
        "&&" => JsonValue.Create(Truthy(l) && Truthy(r)),
        "||" => JsonValue.Create(Truthy(l) || Truthy(r)),
        "in" => JsonValue.Create(InOp(l, r)),
        _ => null,
    };

    static bool InOp(JsonNode? l, JsonNode? r)
    {
        if (r is JsonArray a) return a.Any(x => JsonNode.DeepEquals(x, l));
        if (r is JsonValue sv && sv.TryGetValue<string>(out var s))
            return s.Contains(IrFilters.ToString(l));
        return false;
    }

    static bool AsDouble(JsonNode? n, out double d)
    {
        d = 0;
        return n is JsonValue jv && jv.TryGetValue(out d);
    }

    static bool Truthy(JsonNode? v) => v switch
    {
        null => false,
        JsonValue jv when jv.TryGetValue<bool>(out var b) => b,
        JsonValue jv when jv.TryGetValue<double>(out var d) => d != 0,
        JsonValue jv when jv.TryGetValue<string>(out var s) => s.Length > 0,
        JsonArray a => a.Count > 0,
        JsonObject o => o.Count > 0,
        _ => true,
    };
}

/// <summary>Internal engine handle. Applications should use the public <c>Dejavu</c> package.</summary>
public sealed class EngineHandle
{
    public JsonObject Parse(string source) => T1Parser.ParseToIr(source);
    public string Render(JsonObject ir, JsonObject ctx) => IrRenderer.Render(ir, ctx);
    public string RenderSource(string source, JsonObject ctx) => IrRenderer.RenderSource(source, ctx);
}

File: projects/dejavu.kt/dejavu-engine/src/main/kotlin/dejavu/engine/IrRenderer.kt
package dejavu.engine

import dejavu.language.T1Parser
import dejavu.types.applyFilter
import dejavu.types.valueToString
import kotlinx.serialization.json.*

/**
 * Canonical IR evaluator. Same IR + Context must yield identical output
 * across host languages (byte-for-byte for T1 cases).
 */
object IrRenderer {
    fun render(doc: JsonObject, ctx: JsonObject): String {
        val scope = ctx.toMutableMap()
        return renderNode(doc["body"]!!, scope)
    }

    fun renderSource(source: String, ctx: JsonObject): String =
        render(T1Parser.parseToIr(source), ctx)

    private fun renderNode(node: JsonElement, scope: MutableMap<String, JsonElement>): String {
        val obj = node.jsonObject
        return when (obj["type"]!!.jsonPrimitive.content) {
            "Template" -> obj["children"]!!.jsonArray.joinToString("") { renderNode(it, scope) }
            "Text" -> obj["value"]!!.jsonPrimitive.content
            "Comment" -> ""
            "Interpolation" -> {
                val s = valueToString(eval(obj["expression"]!!, scope))
                if (obj["raw"]?.jsonPrimitive?.booleanOrNull == true) s else htmlEscape(s)
            }

            "Stmt.If" -> renderIf(obj, scope)
            "Stmt.For" -> renderFor(obj, scope)
            "Stmt.Raw" -> obj["value"]!!.jsonPrimitive.content
            "Stmt.Block" -> obj["body"]!!.jsonArray.joinToString("") { renderNode(it, scope) }
            "Stmt.Extends", "Stmt.Include", "Stmt.Super" ->
                error("extends/include/super require a template loader")

            else -> error("node not renderable: ${obj["type"]}")
        }
    }

    private fun renderIf(obj: JsonObject, scope: MutableMap<String, JsonElement>): String {
        if (truthy(eval(obj["test"]!!, scope))) {
            return obj["consequent"]!!.jsonArray.joinToString("") { renderNode(it, scope) }
        }
        for (ei in obj["elseIfs"]!!.jsonArray) {
            val e = ei.jsonObject
            if (truthy(eval(e["test"]!!, scope))) {
                return e["consequent"]!!.jsonArray.joinToString("") { renderNode(it, scope) }
            }
        }
        val alt = obj["alternate"]?.jsonArray ?: return ""
        return alt.joinToString("") { renderNode(it, scope) }
    }

    private fun renderFor(obj: JsonObject, scope: MutableMap<String, JsonElement>): String {
        val item = obj["item"]!!.jsonPrimitive.content
        val iterable = eval(obj["iterable"]!!, scope).jsonArray
        val indexName = obj["index"]?.jsonPrimitive?.content
        val out = StringBuilder()
        iterable.forEachIndexed { i, value ->
            val prev = scope[item]
            scope[item] = value
            val prevIdx = indexName?.let { scope[it] }
            if (indexName != null) scope[indexName] = JsonPrimitive(i)
            out.append(obj["body"]!!.jsonArray.joinToString("") { renderNode(it, scope) })
            if (prev == null) scope.remove(item) else scope[item] = prev
            if (indexName != null) {
                if (prevIdx == null) scope.remove(indexName) else scope[indexName] = prevIdx
            }
        }
        return out.toString()
    }

    private fun eval(expr: JsonElement, scope: Map<String, JsonElement>): JsonElement {
        val obj = expr.jsonObject
        return when (obj["type"]!!.jsonPrimitive.content) {
            "Expr.Literal" -> obj["value"] ?: JsonNull
            "Expr.Identifier" -> scope[obj["name"]!!.jsonPrimitive.content] ?: JsonNull
            "Expr.Member" -> {
                val o = eval(obj["object"]!!, scope)
                if (o is JsonObject) o[obj["property"]!!.jsonPrimitive.content] ?: JsonNull else JsonNull
            }

            "Expr.Index" -> {
                val o = eval(obj["object"]!!, scope)
                val idx = eval(obj["index"]!!, scope)
                when {
                    o is JsonArray && idx is JsonPrimitive && idx.intOrNull != null ->
                        o.getOrNull(idx.int) ?: JsonNull

                    o is JsonObject && idx is JsonPrimitive && idx.isString ->
                        o[idx.content] ?: JsonNull

                    else -> JsonNull
                }
            }

            "Expr.Binary" -> evalBinary(
                obj["operator"]!!.jsonPrimitive.content,
                eval(obj["left"]!!, scope),
                eval(obj["right"]!!, scope),
            )

            "Expr.Unary" -> {
                val v = eval(obj["argument"]!!, scope)
                when (obj["operator"]!!.jsonPrimitive.content) {
                    "!" -> JsonPrimitive(!truthy(v))
                    "-" -> {
                        val n = (v as? JsonPrimitive)?.doubleOrNull
                        if (n != null) JsonPrimitive(-n) else JsonNull
                    }

                    "+" -> v
                    else -> JsonNull
                }
            }

            "Expr.Pipe" -> {
                val value = eval(obj["expression"]!!, scope)
                val args = obj["arguments"]!!.jsonArray.map { eval(it, scope) }
                applyFilter(obj["filter"]!!.jsonPrimitive.content, value, args)
            }

            "Expr.Call" -> error("calls not supported in T1 eval")
            else -> error("invalid expression: ${obj["type"]}")
        }
    }

    private fun evalBinary(op: String, l: JsonElement, r: JsonElement): JsonElement {
        val ld = (l as? JsonPrimitive)?.doubleOrNull
        val rd = (r as? JsonPrimitive)?.doubleOrNull
        return when (op) {
            "+" -> when {
                ld != null && rd != null -> num(ld + rd)
                else -> JsonPrimitive(valueToString(l) + valueToString(r))
            }

            "-" -> if (ld != null && rd != null) num(ld - rd) else JsonNull
            "*" -> if (ld != null && rd != null) num(ld * rd) else JsonNull
            "/" -> if (ld != null && rd != null) num(ld / rd) else JsonNull
            "%" -> if (ld != null && rd != null) num(ld % rd) else JsonNull
            "==" -> JsonPrimitive(l == r)
            "!=" -> JsonPrimitive(l != r)
            "<" -> JsonPrimitive(ld != null && rd != null && ld < rd)
            "<=" -> JsonPrimitive(ld != null && rd != null && ld <= rd)
            ">" -> JsonPrimitive(ld != null && rd != null && ld > rd)
            ">=" -> JsonPrimitive(ld != null && rd != null && ld >= rd)
            "&&" -> JsonPrimitive(truthy(l) && truthy(r))
            "||" -> JsonPrimitive(truthy(l) || truthy(r))
            "in" -> JsonPrimitive(
                when (r) {
                    is JsonArray -> r.any { it == l }
                    is JsonPrimitive -> if (r.isString) r.content.contains(valueToString(l)) else false
                    else -> false
                }
            )

            else -> JsonNull
        }
    }

    private fun num(v: Double): JsonPrimitive =
        if (v % 1.0 == 0.0) JsonPrimitive(v.toLong()) else JsonPrimitive(v)

    private fun truthy(v: JsonElement?): Boolean = when (v) {
        null, JsonNull -> false
        is JsonPrimitive -> when {
            v.booleanOrNull != null -> v.boolean
            v.doubleOrNull != null -> v.double != 0.0
            v.isString -> v.content.isNotEmpty()
            else -> true
        }

        is JsonArray -> v.isNotEmpty()
        is JsonObject -> v.isNotEmpty()
        else -> true
    }

    private fun htmlEscape(s: String): String =
        s.replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace("\"", "&quot;")
            .replace("'", "&#39;")
}

class DejavuEngine {
    fun parse(source: String): JsonObject = T1Parser.parseToIr(source)
    fun render(ir: JsonObject, ctx: JsonObject): String = IrRenderer.render(ir, ctx)
    fun renderSource(source: String, ctx: JsonObject): String = IrRenderer.renderSource(source, ctx)
}
